### ISSUE TITLE
Support creating histograms using logql

### ISSUE THREAD

**Issue by chancez** [grafana/loki#2953]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to create histograms using logql based on an extracted field. We want to track job execution times with histograms on a per-user basis and doing this with labels would have too high of a cardinality.

**Describe the solution you'd like**
The design in the original draft of logql v2 seems reasonable:
https://github.com/cyriltovena/loki/blob/a87cf44829841958a3dd78030266ac93eb7269c3/docs/design-documents/2020-03-logql-fields.md#series--histogram-operators 

**Describe alternatives you've considered**
Using cloudwatch, which has a `bin` function for this. But we're already sending logs to Loki so that's not ideal.

**Additional context**

**Comment by owen-d** [comment]
I think the original histogram doc was largely to enable `quantile_over_time`. For example, the p99 per user could look something like: 
```
quantile_over_time(0.99,
  {cluster="ops-tools1",container="ingress-nginx"}
    | json
    | __error__ = ""
    | unwrap request_time [1m])) by (user)
```

Does that work?

**Comment by chancez** [comment]
We specifically want to specify our buckets. We're trying to measure our job's RPC request duration/latencies during execution and we want to know what percentage of the RPCs fall into specific duration/latency bands. For example, the percentage/number of RPCs that took longer than 5ms, by user.

**Comment by chancez** [comment]
Not stale. After holidays discussion can continue.

**Comment by chancez** [comment]
Not stale.

**Comment by owen-d** [comment]
Keepalive at least until we can get some more feedback during bug scube

**Comment by owen-d** [comment]
Could this look like
```
buckets_over_time([<le1>, <le2>, <le3>], 
  {<selectors>}
  [10m]
)
```

Any chance you'd like to try and implement this feature? I could work you through it, but alternatively I'm not sure when we're going to prioritize this. General consensus is that we'd accept this.

**Comment by chancez** [comment]
I'd potentially be interested. I think logically, I'd start with the parser/lexer and I'd probably just start by looking at how count_over_time is implemented. I can follow up in slack.

**Comment by jcdauchy-moodys** [comment]
@chancez did you find the time to try to code the "buckets_over_time" LogQL function ?

**Comment by jcdauchy-moodys** [comment]
@chancez Did you find the time to try to write something for "buckets_over_time" logQL function ? Thanks

**Comment by chancez** [comment]
Nope, I started on it, but had to focus on other things, and my new role requires approval to do FOSS work, so I'm unlikely to get back to this in the near future.

**Comment by cyriltovena** [comment]
The syntax could be like this, allowing to add as many buckets as needed.
```
buckets_over_time({container="query-frontend"} |= "metrics.go" | logfmt | unwrap duration(duration)[1m],1,10,100) by (type)
```

**Comment by lucas-gonzalez** [comment]
Hi! I ran into this attempting to build a loki based dashboard a short while back. I am open to submitting a pull request _(during the weekend, most likely)_ if nobody is working on this and the suggested syntax still stands.

**Comment by pharaujo** [comment]
For those waiting for this to be implemented, I found a very slow, hacky way of faking `buckets_over_time` by abusing go text templating in `label_format`:

<img width="1151" alt="Screenshot 2022-09-15 at 14 01 47" src="https://user-images.githubusercontent.com/555927/190410539-14bc7dda-cf63-414a-a88f-bdb80ae1f7a4.png">

It'd be great to have a proper way to do this though!

**Comment by mrbrianevans** [comment]
I would like to make a histogram of response times for a webserver. Is that possible with Loki?

(I'm not looking for p99 or anything like that, I just want to see how many requests are in each bucket)

**Comment by pharaujo** [comment]
@mrbrianevans There's currently no function to generate the buckets (that's what this issue is about). That said, that's exactly what I did [in my previous reply](https://github.com/grafana/loki/issues/2953#issuecomment-1248073316). I extracted individual times from the webserver logs, and created a new `le` bucket label where I use go txt/template to conditionally assign bucket values to the log line. Then I do `sum(count_over_time(...)) by (le)` to get the count of requests in each bucket. Do note that doing this is terribly slow and resource-heavy.

**Comment by mrbrianevans** [comment]
The visualisation in your previous post looks like a heatmap of request latency over time. I am looking for simply a histogram of all requests in the currently selected time period. Eg something like this:
![image](https://user-images.githubusercontent.com/53117772/193561645-0ebb0864-6610-4ebb-8409-dba3151eddeb.png)

which shows how many requests fall into each bucket of latency

**Comment by j-karls** [comment]
> For those waiting for this to be implemented, I found a very slow, hacky way of faking `buckets_over_time` by abusing go text templating in `label_format`:
> 
> <img alt="Screenshot 2022-09-15 at 14 01 47" width="1151" src="https://user-images.githubusercontent.com/555927/190410539-14bc7dda-cf63-414a-a88f-bdb80ae1f7a4.png">
> 
> It'd be great to have a proper way to do this though!

Thanks a lot for this, it's a tremendous workaround. I have a small correction for it. 

I had to modify this to ensure I compare integers, rather than strings, in the GO text template boolean operation. Otherwise it will compare the first value of each string, and conclude (with buckets of 10ms,20ms,30ms,40ms,50ms,60ms) that a performance of 221ms belongs into the bucket of "30ms and below", rather than in what I've put in my last else-clause. 

My final query looked something like this:
sum(count_over_time({host="host", container_name="container"} |= "ExecutionTimeAdvice" | json | line_format "{{.message}}" | pattern `<_> Method Name: <method_name>. Execution Time: <exe_time>ms` | label_format le="{{ if le (int .exe_time) 10 }}10{{ else if le (int .exe_time) 20 }}20{{ else if le (int .exe_time) 30 }}30{{ else if le (int .exe_time) 40 }}40{{ else if le (int .exe_time) 50 }}50{{ else }}100{{ end }}" [1h])) by (le)

**Comment by jcdauchy-moodys** [comment]
Is it something that might come some time, now that Loki 3 is out ? 

Thanks

**Comment by joeblubaugh** [comment]
@owen-d could you add some more details about how this feature could be implemented? I'm up for taking a whack at it, as it would be useful for some things I'd like to use Loki for. In particular, I want to generate frontend metrics based on logs in Grafana Faro. If you don't hook up a traces backend to Faro, it's hard to get metrics.

**Comment by awoimbee** [comment]
> I found a very slow, hacky way

Honestly it's not so bad, it's possible to dynamically create the buckets, eg for 20ms buckets with a 200ms ceiling:
```logql
sum(
    count_over_time(
        {app="myapp"}
        |= "duration" | json | line_format "{{.duration}}" | pattern "<exe_time>ms"
        | label_format bucket="{{ if gt (int .exe_time) 200 }}inf{{ else }}{{ div (add (int .exe_time) 19) 20 | mul 20 }}{{ end }}"
        | __error__="" [1h]
    )
) by (bucket)
```
